src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -2436,118 +2477,321 @@ public void UpdateIsNullableEnabledForConstructorsAndFields(bool useStatic, bool
+
+            private ref bool GetIsNullableEnabledForConstructorsAndFields(bool useStatic)
+            {
+                return ref useStatic ? ref IsNullableEnabledForStaticConstructorsAndFields : ref IsNullableEnabledForInstanceConstructorsAndFields;
+            }
 
-                foreach (FieldOrPropertyInitializer.Builder initializer in group)
+            internal static ImmutableArray<ImmutableArray<FieldOrPropertyInitializer>> ToReadOnlyAndFree(ArrayBuilder<ArrayBuilder<FieldOrPropertyInitializer>> initializers)
+            {
+                if (initializers.Count == 0)
                 {
-                    builder.Add(new FieldOrPropertyInitializer(initializer.FieldOpt, initializer.Syntax, syntaxLength));
+                    initializers.Free();
+                    return ImmutableArray<ImmutableArray<FieldOrPropertyInitializer>>.Empty;
+                }
 
-                    // A constant field of type decimal needs a field initializer, so
-                    // check if it is a metadata constant, not just a constant to exclude
-                    // decimals. Other constants do not need field initializers.
-                    if (initializer.FieldOpt == null || !initializer.FieldOpt.IsMetadataConstant)
-                    {
-                        // ignore leading and trailing trivia of the node:
-                        syntaxLength += initializer.Syntax.Span.Length;
-                    }
+                var builder = ArrayBuilder<ImmutableArray<FieldOrPropertyInitializer>>.GetInstance(initializers.Count);
+                foreach (ArrayBuilder<FieldOrPropertyInitializer> group in initializers)
+                {
+                    builder.Add(group.ToImmutableAndFree());
                 }
 
-                group.Free();
+                initializers.Free();
                 return builder.ToImmutableAndFree();
             }
 
             public void Free()
             {
-                NonTypeNonIndexerMembers.Free();
-                StaticInitializers.Free();
-                InstanceInitializers.Free();
-                IndexerDeclarations.Free();
+                NonTypeNonIndexerMembers?.Free();
+                InstanceInitializersForPositionalMembers?.Free();
             }
+        }
 
-            internal void AddOrWrapTupleMembers(SourceMemberContainerTypeSymbol type)
+        protected virtual MembersAndInitializers? BuildMembersAndInitializers(DiagnosticBag diagnostics)
+        {
+            var declaredMembersAndInitializers = getDeclaredMembersAndInitializers();
+            if (declaredMembersAndInitializers is null)
             {
-                this.NonTypeNonIndexerMembers = type.AddOrWrapTupleMembers(this.NonTypeNonIndexerMembers.ToImmutableAndFree());
+                // Another thread completed the work before this one
+                return null;
+            }
+
+            var membersAndInitializersBuilder = new MembersAndInitializersBuilder(declaredMembersAndInitializers);
```

Comment:
>var membersAndInitializersBuilder = new MembersAndInitializersBuilder(declaredMembersAndInitializers); [](start = 12, length = 102)

I think for correctness we should do a volatile check for ```_lazyMembersAndInitializers``` here, before the ```AddSynthesizedMembers``` call. The reason is that, if, due to a race, a loosing thread manages to recalculate ```declaredMembersAndInitializers``` after ```_lazyDeclaredMembersAndInitializers``` has been reset to null, we shouldn't allow it to continue with trying to add synthesized members because that involves binding and might expose transient symbols from ```declaredMembersAndInitializers``` to the outside world. Since the reset to null can happen only after ```_lazyMembersAndInitializers``` is initialized, checking its most recent value should be enough to detect the fact that we lost and shouldn't continue.

Reply:
I think I have a solution using three values: (1) a sentinel value, (2) real result, (3) null.
Then we can distinguish between a transition from 1->2 (should happen at most once, can report diagnostics) and 2->3. 
Also this way we're not holding onto to bad symbols that nobody should look at but could accidentally leak.

---
In reply to: [563007635](https://github.com/dotnet/roslyn/pull/50650#discussion_r563007635) [](ancestors = 563007635)

Reply:
> I think I have a solution using three values: (1) a sentinel value, (2) real result, (3) null.
Then we can distinguish between a transition from 1->2 (should happen at most once, can report diagnostics) and 2->3. 
Also this way we're not holding onto to bad symbols that nobody should look at but could accidentally leak.

Sounds good

---
In reply to: [563014806](https://github.com/dotnet/roslyn/pull/50650#discussion_r563014806) [](ancestors = 563014806,563007635)